packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -456,21 +502,32 @@ export class PendingStateManager implements IDisposable {
 			"No pending message found as we start processing this remote batch",
 		);
 
-		// Note: This could be undefined if this batch became empty on resubmit.
-		// In this case the next pending message is an empty batch marker.
-		// Empty batches became empty on Resubmit, and submit them and track them in case
-		// a different fork of this container also submitted the same batch (and it may not be empty for that fork).
-		const firstMessage = batch.messages.length > 0 ? batch.messages[0] : undefined;
-		const expectedPendingBatchLength = batch.messages.length === 0 ? 1 : batch.messages.length;
+		// If this batch became empty on resubmit, batch.messages will be empty (so firstMessage undefined)
+		// and the next pending message should be an empty batch marker.
+		// More Info: We must submit empty batches and track them in case a different fork
+		// of this container also submitted the same batch (and it may not be empty for that fork).
+		const firstMessage = batch.messages[0];
+		const expectedPendingBatchLength =
+			pendingMessage.batchInfo.length === -1
+				? -1 // Ignore the actual incoming length; -1 length is for back compat so force this to match
+				: batch.messages.length === 0
+					? 1 // For an empty batch, expect a singleton array with the empty batch marker
+					: batch.messages.length; // Otherwise, the lengths should actually match
+
+		// Note: We don't need to use getEffectiveBatchId here, just check the explicit stamped batchID
+		// That logic is needed only when comparing across potential container forks.
+		// Furthermore, we also are comparing the batch IDs constituent data - clientId (it's local) and batchStartCsn.
+		const pendingBatchId = asBatchMetadata(pendingMessage.opMetadata)?.batchId;
+		const inboundBatchId = batch.batchId;
 
 		// We expect the incoming batch to be of the same length, starting at the same clientSequenceNumber,
-		// as the batch we originally submitted.
+		// as the batch we originally submitted. The batchIds should match as well, if set (or neither shoudl be set)
 		// We have another later check to compare the message contents, which we'd expect to fail if this check does,
-		// so we don't throw here, merely log.  In a later release this check may replace that one.
+		// so we don't throw here, merely log.  In a later release this check may replace that one since it's cheaper.
```

Comment:
Is it worth making this change now? I think it would help clean up the code a bit.

Reply:
Since we don't understand what's causing the hits on the other check and I just added that additional logging, I'd like to wait and see what we can learn there before changing anything.  In 2.2 we'll have both that new log and this log (minus the batchId check, which will always be `undefined` for now since this feature isn't used yet), and can see how they relate.